third_party/amd/lib/TritonAMDGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandWMMA.cpp

```diff
@@ -0,0 +1,226 @@
+ * (the "Software"), to deal in the Software without restriction,
+ * including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#ifdef USE_ROCM
+
+#include "../PatternTritonGPUOpToLLVM.h"
+#include "SharedToDotOperandMatrixCore.h"
+#include "Utility.h"
+
+using ::mlir::triton::gpu::AMDWmmaEncodingAttr;
+using ::mlir::triton::gpu::DotOperandEncodingAttr;
+using ::mlir::triton::gpu::getOrder;
+using ::mlir::triton::gpu::getShapePerCTA;
+using ::mlir::triton::gpu::SharedEncodingAttr;
+
+namespace SharedToDotOperandWMMA {
+
+/**
+ * @brief This function maps particular load of wmma dot operand to element
+ * indexes(row, col)
+ *
+ * Whole tensor is broken into "blocks" of warps along "non-K" axis.
+ * One block could be processed by multiple warps.
+ * One warp works on a piece of tensor size elemsPerInstr[0] x K.
+ * Each of these pieces is broken into "tiles" of size elemsPerInstr[0] x
+ * elemsPerInstr[1].
+ *
+ * Total offset of element is a sum of following values:
+ * 1. Offset of warp block in tensor
+ * 2. Offset of warp inside one warp block
+ * 3. Offset of tile in one warp
+ * 4. Offset of one lane data in a tile
+ * 5. Offset of particular element of tensor processed by one lane
+ *
+ * This function computes these offsets for axies independently
+ *
+ * @param rewriter
+ * @param loc
+ * @param elemsPerInstr operand tile shape consumed by one WMMA instruction
+ * @param warpId id component of 2d warp grid along nono-K axis
+ * @param laneId lane id in warp [0..63]
+ * @param warpsPerGroup number of warps in one block
+ * @param numOfElems number of elements accessed by thread per repetition
+ * @param reps number of instructions repretition to fully cover dot operand
+ * @param smemStrides strides in LDS tensor
+ * @param loadVecSize number of elements loaded by one operation
+ * @param iNonKDim non-K dimension of dot operand
+ * @return vector (i-th element corresponds to i-th load instruction) of
+ * 2-element vectors(tensor row and col).
+ */
+llvm::SmallVector<llvm::SmallVector<Value>>
+computeTensorElemMapping(ConversionPatternRewriter &rewriter, Location loc,
```

Comment:
Can we have a better description/name for this functon. Is it mapping the loads that a thread is going to execute to the element indices in the full tensor ? Is there a reason why we can't rely on `emitOffsets`/`emitIndices` instead of calculating the mapping in here ?

Reply:
cc+ @binarman

Reply:
@zahimoud it just occurred to me that we don't have a case for `dotOperandLayout` in `emitOffsetForLayout`, do we?

Reply:
Right, but is the mapping related to the parent layout offsets ? I'm actually not sure what is the smem access pattern for wmma and mfma, but just wondering if the emitOffsets/emitIndices might make the code a bit easier.

Reply:
No, the mapping is about operand A and B tensors. We need to figure out where is the x-th element of thread y in shared memory. So we first compute the coordinates of the element in the original tensor. And then compute the shared memory offsets of the element based on shared memory swizzling pattern.